mod-source-record-storage-server/src/main/java/org/folio/services/impl/LBRecordServiceImpl.java

```diff
@@ -1,15 +1,259 @@
+      .compose(optionalRecord -> optionalRecord
+        .map(r -> {
+          if (nonNull(record.getParsedRecord()) && isEmpty(record.getParsedRecord().getId())) {
+            record.getParsedRecord().setId(record.getId());
+          }
+          if (nonNull(record.getErrorRecord()) && isEmpty(record.getErrorRecord().getId())) {
+            record.getErrorRecord().setId(record.getId());
+          }
+          return insertOrUpdateRecord(connection, record, tenantId);
+        })
+        .orElse(Future.failedFuture(new NotFoundException(
+          format("Record with id '%s' was not found", record.getId())))))
+      );
+  }
+
+  @Override
+  public Future<Record> getFormattedRecord(String externalIdIdentifier, String id, String tenantId) {
+    // NOTE: if this needs optimization, can try SQL functions to get partial record with parsed record and format
+    Future<Optional<Record>> future;
+    if (isNotEmpty(externalIdIdentifier)) {
+      ExternalIdType externalIdType = getExternalIdType(externalIdIdentifier);
+      future = dao.getRecordByExternalId(id, externalIdType, tenantId);
+    } else {
+      future = dao.getById(id, tenantId);
+    }
+    return future.map(record -> record
+      .orElseThrow(() -> new NotFoundException(format("Couldn't find record with %s %s", externalIdIdentifier, id))))
+      .compose(record -> parsedRecordDao.getById(record.getId(), tenantId)
+      .map(parsedRecord -> parsedRecord
+        .orElseThrow(() -> new NotFoundException(format("Couldn't find parsed record with id %s", record.getId()))))
+      .map(parsedRecord -> getRecordWithParsedRecord(record, parsedRecord)));
+  }
+
+  @Override
+  public Future<Record> updateSuppressFromDiscoveryForRecord(SuppressFromDiscoveryDto suppressFromDiscoveryDto, String tenantId) {
+    String id = suppressFromDiscoveryDto.getId();
+    IncomingIdType idType = suppressFromDiscoveryDto.getIncomingIdType();
+    Boolean suppressDiscovery = suppressFromDiscoveryDto.getSuppressFromDiscovery();
+    return dao.inTransaction(tenantId, connection ->
+      dao.getRecordById(connection, id, idType, tenantId)
+        .map(record -> record.orElseThrow(() -> new NotFoundException(format("Couldn't find record with %s %s", idType, id))))
+        .map(record -> record.withAdditionalInfo(record.getAdditionalInfo().withSuppressDiscovery(suppressDiscovery)))
+        .compose(record -> dao.update(connection, record, tenantId)));
+  }
+
+  @Override
+  public Future<Record> updateSourceRecord(ParsedRecordDto parsedRecordDto, String snapshotId, String tenantId) {
+    String id = parsedRecordDto.getId();
+    // no processing of the record is performed apart from the update itself
+    Snapshot snapshot = new Snapshot()
+      .withJobExecutionId(snapshotId)
+      .withStatus(Snapshot.Status.COMMITTED);
+    return dao.inTransaction(tenantId, connection ->
+      rawRecordDao.getById(connection, id, tenantId)
```

Comment:
I'm not sure I understand why we need to search for rawRecord if it is not updated by this operation. And as for the id value of the record, would it match rawRecord id?

Reply:
The new record table id is now a foreign key for raw record table id, marc record table id, and error record table id. This look is performed to get the existing raw record to add it to the new record created. Previously, the record lookup via SQL function with join to all related tables returned the raw record. The save of new record and old record follows previous save logic.

Reply:
There is absolutely room for optimization here, but did not feel it justified as it is only updating single source record.

Reply:
Optimization would require changes in [saveAndUpdate](https://github.com/folio-org/mod-source-record-storage/pull/159/files/22c9fa7e613a93cf02a23086a75bf8e41c786f16#diff-08b1dfa84e6c6cae8956502ce38a64c8R232), [insertOrUpdateRecord](https://github.com/folio-org/mod-source-record-storage/pull/159/files/22c9fa7e613a93cf02a23086a75bf8e41c786f16#diff-08b1dfa84e6c6cae8956502ce38a64c8R195), etc.

Reply:
OK, got it, thanks